aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -410,6 +427,177 @@ impl AptosVM {
         }
     }
 
+    // Execute a multisig transaction:
+    // 1. Obtain the payload of the transaction to execute. This could have been stored on chain
+    // when the multisig transaction was created.
+    // 2. Execute the target payload. If this fails, discard the session and keep the gas meter and
+    // failure object. In case of success, keep the session and also do any necessary module publish
+    // cleanup.
+    // 3. Call post transaction cleanup function in multisig account module with the result from (2)
+    fn execute_multisig_transaction<S: MoveResolverExt + StateView>(
+        &self,
+        storage: &S,
+        mut session: SessionExt<S>,
+        gas_meter: &mut AptosGasMeter,
+        txn_data: &TransactionMetadata,
+        txn_payload: &Multisig,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(VMStatus, TransactionOutputExt), VMStatus> {
+        fail_point!("move_adapter::execute_multisig_transaction", |_| {
+            Err(VMStatus::Error(
+                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+            ))
+        });
+
+        gas_meter
+            .charge_intrinsic_gas_for_transaction(txn_data.transaction_size())
+            .map_err(|e| e.into_vm_status())?;
+
+        // Step 1: Obtain the payload. If any errors happen here, the entire transaction should fail
+        let provided_payload = if txn_payload.transaction_payload.is_some() {
+            bcs::to_bytes(txn_payload.transaction_payload.as_ref().unwrap()).unwrap()
+        } else {
+            // Default to empty bytes if payload is not provided.
+            bcs::to_bytes::<Vec<u8>>(&vec![]).unwrap()
+        };
+        // Failures here will be propagated back.
+        let payload_bytes: Vec<Vec<u8>> = session
+            .execute_function_bypass_visibility(
+                &MULTISIG_ACCOUNT_MODULE,
+                GET_NEXT_TRANSACTION_PAYLOAD,
+                vec![],
+                serialize_values(&vec![
+                    MoveValue::Address(txn_payload.multisig_address),
+                    MoveValue::vector_u8(provided_payload.clone()),
+                ]),
+                gas_meter,
+            )?
+            .return_values
+            .into_iter()
+            .map(|(bytes, _ty)| bytes)
+            .collect::<Vec<_>>();
+        // We have to deserialize once to get the actual return type of the function, which is
+        // vec<u8>.
+        let payload: Vec<u8> = bcs::from_bytes(payload_bytes.first().unwrap()).map_err(|_| {
+            PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined)
+        })?;
+        // Deserialize a second time to convert bytes into the actual multisig payload type.
+        let payload: EntryFunction = bcs::from_bytes(&payload).map_err(|_| {
+            PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined)
```

Comment:
same as above

also you could probably merge 2 or 3 of these into a map call...

payload_bytes.first().map(...)

Reply:
Updated